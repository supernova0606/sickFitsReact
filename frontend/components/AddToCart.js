import { Component } from 'react';
import { graphql, compose } from 'react-apollo';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { removeFromCartEnhancer, userEnhancer, addToCartEnhancer, singleItemEnhancer } from '../enhancers/enhancers';

class AddToCart extends Component {
  static propTypes = {
    currentUser: PropTypes.object.isRequired,
  };

  componentDidMount() {
    this.props.currentUser.refetch();
  }

  handleAddToCart = async () => {
    console.log(`Gonna add this item to the cart! ${this.props.id}`);
    const res = await this.props.addToCart({
      variables: {
        id: this.props.id,
      },
    });
    // this.props.currentUser.refetch();
  };

  render() {
    return <button onClick={this.handleAddToCart}>ðŸ›’ Add To Cart</button>;
  }
}

export default compose(userEnhancer, addToCartEnhancer)(AddToCart);
