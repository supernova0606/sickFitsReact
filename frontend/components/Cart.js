import { Component } from 'react';
import { graphql, compose } from 'react-apollo';
import styled from 'styled-components';
import { CURRENT_USER_QUERY } from '../queries';
import RemoveFromCart from './RemoveFromCart';
import TakeMyMoney from './TakeMyMoney';
import { formatMoney } from '../lib/formatMoney';

const CartStyles = styled.div`
  background: white;
  border-radius: 10px;
  padding: 20px;
  overflow: hidden;
  display: flex;
  align-items: center;
`;

class Cart extends Component {
  componentDidMount() {
    console.log('refetching..');
    setTimeout(this.props.currentUser.refetch, 10);
  }
  render() {
    if (!this.props.currentUser.me) {
      return null;
    }
    // Check for loading state..
    // const { loading, error } = this.props.currentUserQuery;
    // const { user } = this.props.currentUserQuery;
    // if (loading || error || !user) return <p>Cart Loading...</p>;
    // const total = user.cart.reduce((a, b) => a + b.price, 0);

    const { me } = this.props.currentUser;
    console.log(me);
    return (
      <CartStyles className="cart">
        <p>ðŸ›’: {me.cart.length} Items</p>
        <ul>
          {me.cart.map(cartItem => (
            <li key={cartItem.id}>
              <strong>
                {cartItem.quantity} of {cartItem.item.title}
              </strong>
              <RemoveFromCart id={cartItem.id} />
            </li>
          ))}
        </ul>

        <TakeMyMoney>
          <hr />
          <button>Checkout!!!</button>
        </TakeMyMoney>
        {/* There
        {user.cart.length === 1 ? ' is ' : 'are '}
        <ChaChing amount={user.cart.length} />
        {user.cart.length === 1 ? ' item ' : ' items '}
        in your cart totaling */}
        {/* <ChaChing amount={formatMoney(total)} /> */}
      </CartStyles>
    );
  }
}

const userEnhancer = graphql(CURRENT_USER_QUERY, { name: 'currentUser' });
export default compose(userEnhancer)(Cart);
