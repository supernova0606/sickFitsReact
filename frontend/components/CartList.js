import { Component } from 'react';
import withData from '../lib/withData';
import Items from '../components/Items';
import Signup from '../components/Signup';
import LoginAuth0 from '../components/LoginAuth0';
import Page from '../components/Page';
import { USER_ORDERS_QUERY, GET_CART_STATE } from '../queries';
import { graphql, compose } from 'react-apollo';
import has from 'lodash.has';
import get from 'lodash.get';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import makeImage from '../lib/image';
import TakeMyMoney from './TakeMyMoney';
import { removeFromCartEnhancer } from '../enhancers/enhancers';
import { CURRENT_USER_QUERY } from '../queries';
import PropTypes from 'prop-types';

const CartStyles = styled.div`
  padding: 20px;
  position: relative;
  background: white;
  position: fixed;
  height: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  transform: translateX(100%);
  transition: all 0.3s;
  box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.2);
  z-index: 5;
  ${props => props.open && `transform: translateX(0);`};
  .toggleOpen {
    position: absolute;
    top: 0;
    left: 0;
    transform: translateX(-100%);
  }
`;

class CartList extends Component {
  static contextTypes = {
    client: PropTypes.object,
  };

  componentDidMount() {
    setTimeout(this.props.currentUserQuery.refetch, 1);
    // Set the cart state to be closed
    this.context.client.writeQuery({
      query: GET_CART_STATE,
      data: { ui: { isCartOpen: true, __typename: 'Network' } },
    });
  }

  toggleOpen = () => {
    this.props.uiQuery.updateQuery(() => ({ ui: { isCartOpen: false, __typename: 'Network' } }));
  };

  render() {
    if (this.props.loading) {
      return <p>Loading....</p>;
    }

    if (this.props.error) {
      return <p>Error....</p>;
    }

    // if (!has(this.props, 'currentUserQuery.user.cart')) {
    //   return <p>Don't have it yet!</p>;
    // }
    // OMG Clean this up
    let { user = {} } = this.props.currentUserQuery || {};
    if (!user) user = {};
    const cart = user.cart || [];
    const userId = user.id;
    // const userId = this.props.currentUserQuery.user.id;

    const total = cart.reduce((a, b) => a + b.price, 0);
    return (
      <CartStyles open>
        <button className="toggleOpen" onClick={this.toggleOpen}>
          Open Cart
        </button>
        <h1>{cart.length} Items</h1>

        <ul>
          {cart.map(item => (
            <li key={item.id}>
              {item.title}
              <button
                onClick={() =>
                  this.props.removeFromCart({
                    variables: {
                      userId,
                      itemId: item.id,
                    },
                  })}
              >
                &times; Delete
              </button>
            </li>
          ))}
        </ul>
        <TakeMyMoney amount={total} name="Testing 123" description="Test test 123">
          <button>Buy for {formatMoney(total)}</button>
        </TakeMyMoney>
      </CartStyles>
    );
  }
}

const userEnhancer = graphql(CURRENT_USER_QUERY, { name: 'currentUserQuery' });
const uiEnhancer = graphql(GET_CART_STATE, { name: 'uiQuery' });
export default compose(userEnhancer, removeFromCartEnhancer, uiEnhancer)(CartList);
