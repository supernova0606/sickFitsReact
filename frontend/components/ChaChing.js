import TransitionGroup from 'react-transition-group/TransitionGroup';
import Transition from 'react-transition-group/Transition';
import styled from 'styled-components';

const CartCount = styled.div`
  padding: 10px;
  transition: all 0.5s;
  display: inline-block;
  background: white;
  background: linear-gradient(to bottom, #eeeeee 0%, #eeeeee 100%);
  transform: translateY(100%);
  z-index: 1;
  position: relative;
  &.cart-entering {
    background: green;
    transform: translateY(100%);
  }
  &.cart-entered {
    transform: translateY(0);
  }
  &.cart-exiting {
    position: absolute;
    z-index: 0;
    left: 0;
    transform: translateY(-100%);
  }
`;

const ChaChingStyles = styled.div`
  position: relative;
  display: inline-block;
  overflow: hidden;
`;

const ChaChing = ({ amount }) => (
  <ChaChingStyles>
    <TransitionGroup>
      <Transition
        in
        timeout={{
          enter: 0,
          exit: 500,
        }}
        key={amount}
      >
        {status => (
          <CartCount className={`cart-${status}`} key={`count-${amount}`}>
            {amount}
          </CartCount>
        )}
      </Transition>
    </TransitionGroup>
  </ChaChingStyles>
);

export default ChaChing;
