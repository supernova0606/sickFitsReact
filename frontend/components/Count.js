// TODO THis is not needed
import React, { Component } from 'react'
import { graphql, gql } from 'react-apollo'

import { ALL_ITEMS_QUERY } from '../queries';

class Count extends Component {

  render() {

    // 1
    if (this.props.allLinksQuery && this.props.allLinksQuery.loading) {
      return <div>Loading</div>
    }

    // 2
    if (this.props.allLinksQuery && this.props.allLinksQuery.error) {
      return <div>Error</div>
    }

    // 3
    const itemsToRender = this.props.allLinksQuery.allItems

    return (
      <div>
        <h1>There are {itemsToRender.length} items for sale</h1>
      </div>
    )
  }

}

// 1

// We export the graphQL HOC - this will fetch the data and inject it into the Count compeont via props

export default graphql(ALL_ITEMS_QUERY, { name: 'allLinksQuery' })(Count)
