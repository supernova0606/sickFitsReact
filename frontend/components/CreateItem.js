import React, { Component } from 'react';
import { graphql, gql } from 'react-apollo';
import { CREATE_ITEM_MUTATION } from '../queries';
import ErrorMessage from './ErrorMessage';
import Form from './styles/Form';
import PropTypes from 'prop-types';

class CreateItem extends Component {
  static propTypes = {
    createItemMutation: PropTypes.func.isRequired,
  };

  state = {
    item: {
      title: '',
      description: '',
      image: '',
      largeImage: '',
      price: 0,
    },
    loading: false,
    error: {
      message: null,
    },
  };

  handleChange = e => {
    const { name, value, type } = e.target;
    const updatedItem = {
      ...this.state.item,
      [name]: type === 'number' ? parseFloat(value) : value,
    };
    this.setState({ item: updatedItem });
  };

  uploadFile = async e => {
    this.setState({ loading: true });
    const files = e.currentTarget.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');

    // use the file endpoint
    const res = await fetch('https://api.cloudinary.com/v1_1/wesbos/image/upload', {
      method: 'POST',
      body: data,
    });
    const file = await res.json();
    this.setState({ image: file.secure_url, largeImage: file.eager[0].secure_url, loading: false });
  };

  createItem = async e => {
    e.preventDefault();
    // TODO: handle any errors
    // turn loading on
    this.setState({ loading: true });
    try {
      const res = await this.props.createItemMutation({
        // pass in those variables from state
        variables: {
          ...this.state.item,
        },
      });
    } catch (error) {
      this.setState({ error });
    }
    this.setState({ loading: false });
  };

  render() {
    return (
      <div>
        <Form onSubmit={this.createItem}>
          {this.state.loading ? 'LOADING...' : null}
          <ErrorMessage error={this.state.error} onButtonClick={() => this.setState({ error: {} })} />
          <p>
            Image
            <input onChange={this.uploadFile} type="file" accept=".png, .jpg, .jpeg" />
            {this.state.image ? <img src={this.state.image} width="100" alt={this.state.title} /> : null}
          </p>
          <p>
            Title
            <input
              value={this.state.title}
              onChange={this.handleChange}
              type="text"
              name="title"
              id="title"
              placeholder="Title"
            />
          </p>
          <label>
            Price
            <input
              type="number"
              id="price"
              name="price"
              min="0"
              value={this.state.price}
              onChange={this.handleChange}
            />
          </label>
          <textarea
            id="description"
            name="description"
            value={this.state.description}
            onChange={this.handleChange}
            placeholder="The desc for this item"
          />
          <button disabled={this.state.loading} type="submit">
            Submit
          </button>
        </Form>
      </div>
    );
  }
}

export default graphql(CREATE_ITEM_MUTATION, {
  name: 'createItemMutation',
  options: {
    refetchQueries: ['AllItemsQuery'],
  },
})(CreateItem);

export { CreateItem };
