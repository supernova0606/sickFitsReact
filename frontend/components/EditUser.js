import React from 'react';
import { compose } from 'react-apollo';
import PropTypes from 'prop-types';
import Form from './styles/Form';
import { userEnhancer, updateUserEnhancer } from '../enhancers/enhancers';

class EditUser extends React.Component {
  static propTypes = {
    currentUser: PropTypes.object.isRequired,
    updateUser: PropTypes.func.isRequired,
  };

  state = {
    changes: {},
  };

  handleChange = e => {
    const { name, value } = e.target;
    const { me } = this.props.currentUser;
    const changes = {
      ...this.state.changes,
      [name]: value,
    };
    if (value === me[e.target.name]) {
      delete changes[name];
    }
    this.setState({ changes });
  };

  handleSubmit = async e => {
    e.preventDefault();
    await this.props.updateUser({
      variables: this.state.changes,
    });
    this.props.currentUser.refetch();
    this.setState({ changes: {} });
  };

  render() {
    const { me } = this.props.currentUser;
    if (!me) return <p>You must be logged in</p>;
    return (
      <Form onSubmit={this.handleSubmit}>
        <label htmlFor="name">
          Name:
          <input type="text" name="name" defaultValue={me.name} onChange={this.handleChange} />
        </label>
        <button type="submit">Update</button>
        <strong>me:</strong>
        <pre>{JSON.stringify(me.name)}</pre>
        <strong>Change:</strong>
        <pre data-test="change">{JSON.stringify(this.state.changes)}</pre>
      </Form>
    );
  }
}

export default compose(userEnhancer, updateUserEnhancer)(EditUser);
export { EditUser };
