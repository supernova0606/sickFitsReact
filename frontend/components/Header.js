import { Component } from 'react';
import { graphql, compose } from 'react-apollo';
import { CURRENT_USER_QUERY } from '../queries';
import Cart from './Cart';
import Login from './LoginAuth0';
import Search from './Search';
import NProgress from 'nprogress';
import Router from 'next/router';

Router.onRouteChangeStart = url => {
  console.log(`Loading: ${url}`);
  NProgress.start();
};
Router.onRouteChangeComplete = () => NProgress.done();
Router.onRouteChangeError = () => NProgress.done();

class Header extends Component {
  componentDidMount() {
    // This fetches the new data, but doesn't populate the user via props
    // this.props.currentUserQuery.refetch();
    // This fetches the new data, and populates the user via props
    // setTimeout(this.props.currentUserQuery.refetch, 1);
  }

  render() {
    // const user = this.props.currentUserQuery.user || {};
    // const { email = '' } = user;
    return (
      <div>
        {/* <p>
          Signed in as <strong>{email}</strong>
        </p> */}
        {/* <Cart /> */}
        <Search />
      </div>
    );
  }
}

const userEnhancer = graphql(CURRENT_USER_QUERY, { name: 'currentUserQuery' });
// export default compose(userEnhancer)(Header);
export default Header;
