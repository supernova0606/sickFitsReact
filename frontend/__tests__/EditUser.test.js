import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { EditUser } from '../components/EditUser';

const wait = amount => new Promise(resolve => setTimeout(resolve, amount));

const currentUser = {
  me: {
    name: 'Wes Bos',
  },
  refetch() {},
};

describe('<EditUser/>', () => {
  it('renders', () => {
    const wrapper = shallow(<EditUser currentUser={currentUser} updateUser={() => {}} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('asks you to log in when you arent logged in', () => {
    const wrapper = shallow(<EditUser currentUser={{}} updateUser={() => {}} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('displays changes', () => {
    const wrapper = shallow(<EditUser currentUser={currentUser} updateUser={() => {}} />);
    const nameInput = wrapper.find('[name="name"]');
    nameInput.simulate('change', { target: { name: 'name', value: 'Scott' } });
    const diff = wrapper.find('pre[data-test="change"]');
    expect(diff.text()).toBe('{"name":"Scott"}');
  });

  it('calls update user when form is submitted', () => {
    const updateUserSpy = jest.fn();
    const wrapper = shallow(<EditUser currentUser={currentUser} updateUser={updateUserSpy} />);
    const nameInput = wrapper.find('[name="name"]');
    nameInput.simulate('change', { target: { name: 'name', value: 'Scott' } });
    wrapper.simulate('submit', { preventDefault() {} });
    expect(updateUserSpy).toHaveBeenCalledWith({ variables: { name: 'Scott' } });
  });

  it('refetches the current user after an update', async () => {
    const currentUserWithSpy = {
      me: { name: 'wes' },
      refetch: jest.fn(),
    };
    const wrapper = shallow(<EditUser currentUser={currentUserWithSpy} updateUser={() => {}} />);
    wrapper.simulate('submit', { preventDefault() {} });
    await wait(0);
    expect(currentUserWithSpy.refetch).toHaveBeenCalled();
  });
});
