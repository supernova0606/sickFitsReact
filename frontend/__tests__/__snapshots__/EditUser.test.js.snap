// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditUser/> asks you to log in when you arent logged in 1`] = `
<p>
  You must be logged in
</p>
`;

exports[`<EditUser/> renders 1`] = `
<Form
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    Name:
    <input
      defaultValue="Wes Bos"
      name="name"
      onChange={[Function]}
      type="text"
    />
  </label>
  <button
    type="submit"
  >
    Update
  </button>
  <strong>
    me:
  </strong>
  <pre>
    "Wes Bos"
  </pre>
  <strong>
    Change:
  </strong>
  <pre
    data-test="change"
  >
    {}
  </pre>
</Form>
`;
